### ISSUE TITLE
EdgeTPU Runtime dependent on AVX/SSE instructions?

### ISSUE THREAD

**Issue by lawrence-jeff** [google-coral/edgetpu#163]
I have an older AMD64 server system I am trying to use the coral USB accelerator with. I have compiled full tensorflow 2.1 on it and can successfully run tensorflow lite operations (albeit slowly)
I notice when I add the delegate code in, I get the dreaded **Illegal instruction (core dumped)** which with tensorflow means it has compiled optimizations that require a newer CPU - since the tensorflow package I am using I compiled and doesn't I'm assuming this is coming from the edgeTPU runtime?
The requirements don't call this need out: https://coral.ai/docs/accelerator/get-started/#requirements, so hoping that is not the case.

Is there something I might be missing? The exact same code on a newer system works correctly (at least when using the CPU model, I see the other issue related to why it fails when I try to use the coral model outside of the tflite_runtime)

The runtime as I understand it is not open source, is there any way to get around this limitation?

**Comment by Namburger** [comment]
Apologies for the delayed in response, we were all OOO for the holiday!

Anyhow, I saw your suggestions on https://github.com/google-coral/edgetpu/issues/126#issuecomment-654320097 and have forwarded it to our documentation team. 

We had a little "silence" release of [libedgetpu](https://github.com/google-coral/libedgetpu) runtime (we are planning on announcing+releasing more tools this month), so in theory, you could:
1) build a cutting edge libedgetpu with `make` from that repo
2) build your tensorflow package on top of the [same commit](https://github.com/google-coral/libedgetpu/blob/master/bazel/WORKSPACE#L39) that you've built libedgetpu

This have been working very well for me with c++ code (here is a [quick guide](https://github.com/Namburger/edgetpu-minimal-example/tree/custom-tensorflow-build), I have not had the bandwidth to test this in python but it should works. What make our `tflite_runtime` package works is essentially because we used the same exact tensorflow commit that we used to build libedgetpu to build (I hope that sentence wasn't too confusing :) ). Anyhow, if you ends up giving it a try, let me know how it works out for you, I can potentially give it a go once getting some more bandwidths!

**Comment by Namburger** [comment]
FYI, you can also build the `tflite_runtime` package: https://github.com/tensorflow/tensorflow/tree/master/tensorflow/lite/tools/pip_package
but this is most likely not what you're looking to do

**Comment by lawrence-jeff** [comment]
Thanks @Namburger . Sorry for the slow response was tied up this week. I did some further testing and switched to using the tflite runtime (custom compiled with no AVX). I get similar results that leads me to still think its the edgetpu code. Unfortunately my compile skills aren't good enough to figure out how to compile that piece.
Here is why I think there are instructions in the edgetpu code causing issues

Collected two devices (7'th gen i7) and my old Opteron based server - installed clean Ubuntu 18.04 on both
Compiled the tflite runtime 2.3 version on the low end system in a docker image
Installed the custom compiled runtime on both systems
Installed the EdgeTPU runtime (current release version) on both systems
Installed the samples https://github.com/google-coral/tflite/tree/master/python/examples/classification on both
So at this point they have the same software.

Keeping the use case very simple so calling the coral delegate but using the standard model so most work is done on the CPU.

**On the i7**
Ran the parrot test with no delegate passing the non coral model -> Passed 200ms
Added the delegate but still used the non coral model -> Passed 200ms 

**On the old machine**
Ran the parrot test with no delegate passing the non coral model -> Passed 500ms
Added the delegate but still used the non coral model -> Failed with Illegal instruction (Core Dumped)

The behavior I expected is this would also work and have the same time of 500ms since the work is passed to the CPU.

In both cases the two python packages were the same, the OS is the same, the Coral is the same. The custom runtime code is proven to not have an issue since it can run fine (although slow) without the delegate... so the only code left in the picture that  seem to be left is the edgetpu runtime...

I will continue to try to see if I can compile the tpu runtime .. but think it might be beyond me without further instructions.

**Comment by Namburger** [comment]
@lawrence-jeff 
Sorry for the late response, I've also been a little busy :/
It does sounds very convincing, but I didn't expect that, I'll try to confirm this today.

Could you submit the full stacktrace when it crashes? 
```
strace python3 your_python_script.py --args2> strace.txt
```

Also, are there any other logs besides the Core Dumped?

**Comment by lawrence-jeff** [comment]
Here is a stack trace
To keep things simple this is back to the production libedgetpu1-std (14.1)
And tensorflow lite (https://dl.google.com/coral/python/tflite_runtime-2.1.0.post1-cp36-cp36m-linux_x86_64.whl). So basically following the quick start for the Coral USB accelerator. 
[tracefileProd.txt](https://github.com/google-coral/edgetpu/files/4951221/tracefileProd.txt)

**Comment by Namburger** [comment]
@lawrence-jeff hi! We're looking into this, but looking through the stacktrace, libedgetpu was defenitely loaded and communications were established with the accelerator:
```
openat(AT_FDCWD, "/usr/lib/x86_64-linux-gnu/libedgetpu.so.1", O_RDONLY|O_CLOEXEC) = 3
.....
access("/sys/devices/pci0000:00/0000:00:03.2/usb1/1-4/1-4:1.0/uevent", F_OK) = 0
openat(AT_FDCWD, "/sys/devices/pci0000:00/0000:00:03.2/usb1/1-4/1-4:1.0/uevent", O_RDONLY|O_CLOEXEC) = 8
```
There was definitely an illegal instruction that killed the process so it must be true that our library wanted to do something "illegal" with you cpu :)

I started communicating this issue internally and will try to confirm this

**Comment by lawrence-jeff** [comment]
Yes - you definitely get a different error if the accelerator is not attached (and that error is consistent between working systems and non working).

**Comment by Namburger** [comment]
@lawrence-jeff yeah, that's expected :)
could you issue this: 
```
ulimit -c unlimited
```
and then try tracing it again?

Also:
I'm not sure if this will be much different from the strace logs I already asked for, but can you also provide the python trace?
```
python3 -m trace --trace your_python_script.py --args > python_strace.txt
```

**Comment by lawrence-jeff** [comment]
[pythonstacktrace.zip](https://github.com/google-coral/edgetpu/files/4955459/pythonstacktrace.zip)
[tracefileProd-Unlim.txt](https://github.com/google-coral/edgetpu/files/4955462/tracefileProd-Unlim.txt)

**Comment by Namburger** [comment]
@lawrence-jeff is there a core.<process-id> file some where in the directory where you execute the script also?

**Comment by lawrence-jeff** [comment]
@Namburger  There is not, no files other than the ones i explicitly create with the tracing commands. Would it be somewhere else in the system?

**Comment by lawrence-jeff** [comment]
[core.zip](https://github.com/google-coral/edgetpu/files/4955795/core.zip)

**Comment by Namburger** [comment]
@lawrence-jeff So we're not convinced that it is the AVX instruction set specifically, but it is definitely due to us compiling the library on a newer cpu.
```
gdb /usr/lib/x86_64-linux-gnu/libedgetpu.so.1 core
.......
warning: Unable to find libthread_db matching inferior's thread library, thread debugging will not be available.
Core was generated by `python3 classify_imageTFLITE.py --model mobilenet_v2_1.0_224_inat_bird_quant_ed'.
Program terminated with signal SIGILL, Illegal instruction.
#0  0x00007f456c64fc84 in ?? () from /usr/lib/x86_64-linux-gnu/libedgetpu.so.1
[Current thread is 1 (LWP 2948)]
(gdb) bt
#0  0x00007f456c64fc84 in ?? () from /usr/lib/x86_64-linux-gnu/libedgetpu.so.1
#1  0x00007f456c6509a0 in ?? () from /usr/lib/x86_64-linux-gnu/libedgetpu.so.1
#2  0x00007f456c61c096 in ?? () from /usr/lib/x86_64-linux-gnu/libedgetpu.so.1
#3  0x00007f456c6800da in ?? () from /usr/lib/x86_64-linux-gnu/libedgetpu.so.1
#4  0x00007f456c679531 in ?? () from /usr/lib/x86_64-linux-gnu/libedgetpu.so.1
#5  0x00007f456c670df9 in ?? () from /usr/lib/x86_64-linux-gnu/libedgetpu.so.1
#6  0x00007f456c6734c5 in ?? () from /usr/lib/x86_64-linux-gnu/libedgetpu.so.1
#7  0x00007f456c673780 in ?? () from /usr/lib/x86_64-linux-gnu/libedgetpu.so.1
#8  0x00007f456c66f24d in ?? () from /usr/lib/x86_64-linux-gnu/libedgetpu.so.1
#9  0x00007f455a710dae in ?? ()
#10 0x000000000000000e in ?? ()
#11 0x00007ffcc268f550 in ?? ()
#12 0x00007ffcc268f4e0 in ?? ()
#13 0x00007f455a71071f in ?? ()
#14 0x0000000000000004 in ?? ()
#15 0x0000000000000000 in ?? ()
(gdb) r
Starting program: /usr/lib/x86_64-linux-gnu/libedgetpu.so.1 
/bin/bash: /usr/lib/x86_64-linux-gnu/libedgetpu.so.1: Permission denied
/bin/bash: line 0: exec: /usr/lib/x86_64-linux-gnu/libedgetpu.so.1: cannot execute: Permission denied
During startup program exited with code 126.
```

Unfortunately, I'm not sure if there's a way for us to fix this, maybe you can compile `libedgetpu.so` on that machine? 
The library is here: https://github.com/google-coral/libedgetpu

Make sure you change the tensorflow commit from [here](https://github.com/google-coral/libedgetpu/blob/master/bazel/WORKSPACE#L39)
to [this one](https://github.com/google-coral/edgetpu/blob/master/WORKSPACE#L5)
So that it would be compatible with the current release of `tflite_runtime` package

**Comment by lawrence-jeff** [comment]
Shouldn't code that is intended to offload work to an external device be compiled in a 'lowest common denominator' way? Kind of by definition customers might be trying to accelerate a device that can't naively do the processing quickly itself. Does the libedgetpu code need to be particularly optimized to offload work to either the device or the tflite runtime?

At the very least the need for certain hardware should be documented in the requirements linked in my report of the issue.. The CPU I'm using is not that old, its from 2012.. (8 core 32GB system)

I'm happy to try to compile it myself - I gave it a shot earlier but wasn't sure how to get that into an installable python library. Any quick pointers there?

**Comment by Namburger** [comment]
@lawrence-jeff 
> Shouldn't code that is intended to offload work to an external device be compiled in a 'lowest common denominator' way?

I agree fully, but this is definitely not an issue we saw coming. Issues like this is very hard to make decision on, for instance, Ubuntu16.04 recently reach EOL but we still have users using it. We really want to build the library for those machines, but to do that, we'll be building on an older glibc version, which means we'll lose some newer c++17 features. That's why at the end of the day, we decided to open source the library. In this case, I'm not 100% sure if the older CPU is the problem or not, but if it's true, we simply don't have the HW to build the library. I'll definitely try to get this documented, but I just wanted to make sure that it is the issue first. Our apologies.

> I'm happy to try to compile it myself - I gave it a shot earlier but wasn't sure how to get that into an installable python library. Any quick pointers there?

Basically `libedgetpu.so` is just a share object file that `load_delegate` is calling during runtime. First I would remove the old .so library to avoid that being loaded by accident. Then you can build the library [here](https://github.com/google-coral/libedgetpu) like this:
```
$ DOCKER_CPUS="k8" DOCKER_TARGETS=libedgetpu make docker-build
```
The output binary should be in `libedgetpu_project_source/out/k8/...`, there will be a throttled(std) or direct(max) version, you can just move those binaries to `/usr/lib/x86_64-linux-gnu` directory and your python script should load it during runtime!

**Comment by lawrence-jeff** [comment]
Thanks for the extra info, I will give it a shot

**Comment by lawrence-jeff** [comment]
Ok - was able to compile the files on my server and replace them in the directory you specified. Unfortunately no change, same error. I do however see while it compiles lots of references to -msse4.2 . SSE 4.2 is not supported by my CPU (opteron k8). At a glance I don't see where this is coming from but my plan is to try to track that down and remove

**Comment by Namburger** [comment]
@lawrence-jeff sorry for all these toubles, wonder if you can get more info with the debug version of libedgetpu? 
I believe you can add `COMPILATION_MODE=dbg` to [this line](https://github.com/google-coral/libedgetpu/blob/master/Makefile#L134)
to:
```
make CPU=\$${cpu} COMPILATION_MODE=dbg -C $(DOCKER_CONTAINER_WORKSPACE) $(DOCKER_TARGETS) || exit 1; \
```
and then compile again with the command I gave from above.

If you're not using docker, it should be:
```
make COMPILATION_MODE=dbg
```

**Comment by parhedberg** [comment]
I am having this same problem. Do I really need a new host with avx support?

**Comment by pdecat** [comment]
FWIW, I had to patch the coral_crosstool Bazel toolchain to be able to compile libedgetpu on old CPUs without SSE4.2: https://github.com/blakeblackshear/frigate-hass-addons/pull/12/files#diff-ad233d86ed208369892f93711e1f006d9e5ce1d53ba85251259c187b21931e63R24-R29

```
RUN git clone https://github.com/google-coral/libedgetpu.git
RUN cd libedgetpu && git checkout 14eee1a076aa1af7ec1ae3c752be79ae2604a708
RUN echo "deb [arch=amd64] https://storage.googleapis.com/bazel-apt stable jdk1.8" > /etc/apt/sources.list.d/bazel.list \
    && curl -s https://storage.googleapis.com/bazel-apt/doc/apt-key.pub.gpg | apt-key add - \
    && apt-get -qq update \
    && apt-get -qq install bazel=3.2.0
RUN cd /libedgetpu && bazel fetch --experimental_repo_remote_exec @coral_crosstool//...
RUN sed -i 's/"-msse4.2",//g' $HOME/.cache/bazel/*/*/external/coral_crosstool/cc_toolchain_config.bzl.tpl
RUN printf "build --copt=-march=native\nbuild --host_copt=-march=native\n" > ~/.bazelrc
RUN cd /libedgetpu && make libedgetpu
RUN apt-get -qq install debhelper
RUN cd /libedgetpu && make deb
```

It would be awesome if this could be passed as a parameter instead of having to patch the toolchain. Or did I miss something?

**Comment by TonyTromp** [comment]
Just started playing with Coral TPU on an older machine (x86) without AVX
Modified the Dockerfile so it installed a compatible version.

```
# /tutorials-master/docker/classification/Dockerfile

#FROM tensorflow/tensorflow:1.15.5
FROM python:3.9.16-slim-buster

# Install wget (to make life easier below) and editors (to allow people to edit
# the files inside the container)
RUN apt-get update && \
    apt-get install -y wget vim emacs nano git

RUN pip install https://tf.novaal.de/westmere/tensorflow-2.7.0-cp39-cp39-linux_x86_64.whl

# Get the tensorflow models research directory
RUN git clone https://github.com/tensorflow/tensorflow.git --depth 1
#RUN git clone https://github.com/tensorflow/models.git && \
#    (cd models && git checkout f788046ca876a8820e05b0b48c1fc2e16b0955bc) && \
#    mv models /tensorflow/models

... rest of file

```
